% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/splitBAM.R
\name{splitBAM}
\alias{splitBAM}
\title{Split a bam file by given criteria}
\usage{
splitBAM(bamFile, splitby = "strand", chrnameList = NULL, flagList = NULL,
  mapqList = NULL, outfile_prefix, nthreads = 20)
}
\arguments{
\item{bamFile}{Bam file path}

\item{splitby}{How to split the file ("strand","chr","flag","mapq")}

\item{chrnameList}{list of Chromosomes to split by}

\item{flagList}{List of samflags to split by}

\item{mapqList}{List of mapq cutoffs to split by (reads with mapq greater than cutoff will be kept)}

\item{outfile_prefix}{prefix for output filename}

\item{nthreads}{Number of threads to use}
}
\value{
Splitted, sorted and indexed bam files
}
\description{
Split a bam file by given criteria
}
\examples{
bam <- system.file("extdata", "test_2L2R.bam", package="vivlib")
splitBAM(bam = bam, splitby = "chr", chrnameList = NULL, flagList = NULL,
             mapqList = NULL, outfile_prefix = "test", nthreads = 20)

}

